### PR TITLE
feat(apps): create the code of conduct managing functions

### PR DESCRIPTION
**Pull request by josephperrott**
Create the Firebase functions used for managing the code of conduct application.

A follow up PR will contain the frontend application

### CODE REVIEW

File: apps/functions/code-of-conduct/shared.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import * as admin from 'firebase-admin';
+import {Octokit} from '@octokit/rest';
+import {createAppAuth} from '@octokit/auth-app';
+import * as functions from 'firebase-functions';
+
+/** Parameters for blocking a user. */
+export interface BlockUserParams {
+  username: string;
+  blockUntil: Date;
```
Comment (devversion) [inline apps/functions/code-of-conduct/shared.ts:9]: Isn't `blockUntil` delivered as a string to the functions? in e.g. ISO?

File: apps/functions/code-of-conduct/shared.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+/** A CallableContext which is confirmed to already have an authorized user. */
+interface AuthenticatedCallableContext extends functions.https.CallableContext {
+  auth: NonNullable<functions.https.CallableContext['auth']>;
+}
+
+/** Verify that the incoming request is authenticated and authorized for access. */
+export function checkAuthenticationAndAccess(
+  context: functions.https.CallableContext,
+): asserts context is AuthenticatedCallableContext {
+  if (context.auth == undefined) {
```
Comment (devversion) [inline apps/functions/code-of-conduct/shared.ts:61]: We should note how the authentication is done here?

File: apps/functions/code-of-conduct/shared.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    // Throwing an HttpsError so that the client gets the error details.
+    throw new functions.https.HttpsError('unauthenticated', 'This action requires authentication');
+  }
+}
+
+/** Retrieves a Github client instance authenticated as the Angular Robot Github App. */
+export async function getAuthenticatedGithubClient() {
+  const GITHUB_APP_PEM = Buffer.from(process.env['GITHUB_PEM']!, 'base64').toString('utf-8');
```
Comment (devversion) [inline apps/functions/code-of-conduct/shared.ts:69]: I wish we would have named `GITHUB_PEM` instead `GITHUB_APP_PRIVATE_KEY`. The current one seems quite coarse/ambiguous IMO

File: apps/functions/code-of-conduct/syncUsers.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+export const dailySync = functions
+  .runWith({
+    secrets: ['GITHUB_PEM', 'GITHUB_APP_ID'],
+  })
+  .pubsub.schedule('every day 08:00')
+  .timeZone('America/Los_Angeles')
+  .onRun(syncUsers);
+
+export const syncUsersFromGithub = functions
```
Comment (devversion) [inline apps/functions/code-of-conduct/syncUsers.ts:16]: Nit: Add a comment/JSDoc for every action and what it's supposed to do

**Review by devversion — COMMENTED** [review]
overall LGTM. Nice functions!

File: apps/functions/code-of-conduct/shared.ts (hunk 3 above)
Comment (josephperrott) [reply apps/functions/code-of-conduct/shared.ts:69]: Updated.

File: apps/functions/BUILD.bazel [hunk 5]
```diff
@@ -17,13 +17,14 @@ ts_library(
...
     deps = [
+        "//apps/functions/code-of-conduct",
         "//apps/functions/githubWebhook",
         "//apps/functions/ng-dev",
         "@npm//firebase-admin",
     ],
 )
 
-esbuild_esm_bundle(
+esbuild_cjs_bundle(
```
[the remark this answers is not in the document]
Comment (devversion) [reply apps/functions/BUILD.bazel:27]: is there an issue we can link to?

[the remark this answers is not in the document]
Comment (josephperrott) [reply apps/functions/BUILD.bazel:27]: I created this real quick: https://github.com/googleapis/gax-nodejs/issues/1443

[the remark this answers is not in the document]
Comment (devversion) [reply apps/functions/BUILD.bazel:27]: Thanks. I think `firebase-admin` should be able to use it as a CommonJS package using the NodeJS CJS/ESM interop- but I don't know what errors you were seeing in concrete. We can go with CJS for now.

**Review by devversion — APPROVED** [review]

**Pull request closed without merging**